### ISSUE TITLE
Fix module mypy static type check errors

### ISSUE THREAD

**Issue by tahouse** [rnag/dataclass-wizard#51]
* Dataclass Wizard version: 0.22.0
* Python version: 3.7.0, various
* Operating System: CentOS 7

### Description

When using `mypy` to type check a project that imports dataclass-wizard, `mypy` complains:
```
error: Skipping analyzing "dataclass_wizard":
module is installed, but missing library stubs or py.typed marker  [import]
    from dataclass_wizard import DumpMeta, JSONFileWizard, JSONWizard, jso...
    ^
note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
```

Users must use a `#type:ignore` after the import to prevent this error, but this disables type checking for usages of the module. I notice there is good use of type annotations in the codebase, so the work may not end up being too hard to remove this limitation. Note, work on some of https://github.com/rnag/dataclass-wizard/issues/50 may intersect with this issue.

### What I Did
Cloned dataclass-wizard repo, created new conda environment and installed requirements-dev.txt. Then installed mypy `pip install mypy==0.950` and ran `mypy` on source, but mypy fails with many errors. I believe these would need to be fixed to allow mypy static type checking. 

I thought some of these seem likely caused by running in Python 3.7 context, but I tried python 3.8, 3.9, and 3.10 environments and all failed with mostly similar errors. 
```
❯ mypy dataclass_wizard
dataclass_wizard/errors.py:68: error: Incompatible types in assignment (expression has type "str", variable has type "None")
dataclass_wizard/errors.py:139: error: Attribute function "default_factory" with type "Callable[[], _T]" does not accept self argument
dataclass_wizard/utils/string_conv.py:255: error: Need type annotation for "indices"
dataclass_wizard/type_def.py:104: error: Name "TypedDict" is not defined
dataclass_wizard/type_def.py:122: error: Argument 1 to "append" of "list" has incompatible type "object"; expected "Type[Any]"
dataclass_wizard/type_def.py:128: error: Argument 1 to "append" of "list" has incompatible type "object"; expected "Type[Any]"
dataclass_wizard/type_def.py:132: error: Incompatible import of "PyLiteral" (imported name has type "typing_extensions._SpecialForm", local name has type "typing._SpecialForm")
dataclass_wizard/type_def.py:133: error: Incompatible import of "PyProtocol" (imported name has type "typing_extensions._SpecialForm", local name has type "typing._SpecialForm")
dataclass_wizard/type_def.py:140: error: Argument 1 to "append" of "list" has incompatible type "object"; expected "Type[Any]"
dataclass_wizard/type_def.py:144: error: Module "typing" has no attribute "_ForwardRef"; maybe "ForwardRef"?
dataclass_wizard/type_def.py:144: error: Name "PyForwardRef" already defined (possibly by an import)
dataclass_wizard/type_def.py:148: error: Cannot assign to a method
dataclass_wizard/type_def.py:154: error: Module has no attribute "_type_check"; maybe "no_type_check" or "type_check_only"?
dataclass_wizard/type_def.py:162: error: Module has no attribute "_type_check"; maybe "no_type_check" or "type_check_only"?
dataclass_wizard/bases.py:13: error: Invalid assignment target
dataclass_wizard/bases.py:14: error: Type variable "dataclass_wizard.bases.M" is invalid as target for type alias
dataclass_wizard/bases.py:27: error: Self argument missing for a non-static method (or an invalid type for self)
dataclass_wizard/bases.py:62: error: "M" has no attribute "__name__"
dataclass_wizard/bases.py:71: error: "M" has no attribute "__mro__"
dataclass_wizard/bases.py:74: error: Incompatible return value type (got "type", expected "M")
dataclass_wizard/bases.py:74: error: Argument 2 to "type" has incompatible type "Tuple[M]"; expected "Tuple[type, ...]"
dataclass_wizard/bases.py:76: error: Self argument missing for a non-static method (or an invalid type for self)
dataclass_wizard/bases.py:155: error: Incompatible types in assignment (expression has type "None", variable has type "Dict[str, str]")
dataclass_wizard/bases.py:159: error: Incompatible types in assignment (expression has type "None", variable has type "Union[DateTimeTo, str]")
dataclass_wizard/bases.py:166: error: Incompatible types in assignment (expression has type "None", variable has type "Union[LetterCase, str]")
dataclass_wizard/bases.py:173: error: Incompatible types in assignment (expression has type "None", variable has type "Union[LetterCase, str]")
dataclass_wizard/bases.py:183: error: Incompatible types in assignment (expression has type "None", variable has type "str")
dataclass_wizard/bases.py:239: error: Incompatible types in assignment (expression has type "None", variable has type "Dict[Type[Any], Callable[..., Any]]")
dataclass_wizard/bases.py:263: error: Incompatible types in assignment (expression has type "None", variable has type "Dict[Type[Any], Callable[..., Any]]")
dataclass_wizard/utils/typing_compat.py:33: error: Variable "dataclass_wizard.utils.typing_compat.PyTypedDict" is not valid as a type
dataclass_wizard/utils/typing_compat.py:33: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
dataclass_wizard/utils/typing_compat.py:33: error: Invalid base class "PyTypedDict"
dataclass_wizard/utils/typing_compat.py:86: error: Module "typing" has no attribute "_AnnotatedAlias"
dataclass_wizard/utils/typing_compat.py:86: error: Name "_AnnotatedAlias" already defined (possibly by an import)
dataclass_wizard/utils/typing_compat.py:134: error: Module has no attribute "_GenericAlias"
dataclass_wizard/utils/typing_compat.py:136: error: Module has no attribute "GenericAlias"
dataclass_wizard/utils/typing_compat.py:137: error: Module has no attribute "UnionType"; maybe "FunctionType"?
dataclass_wizard/utils/typing_compat.py:160: error: Incompatible types in assignment (expression has type "Tuple[Any, Type[_SpecialForm]]", variable has type "Tuple[Any, Type[_SpecialForm], Any, Any]")
dataclass_wizard/utils/typing_compat.py:161: error: Module has no attribute "_GenericAlias"
dataclass_wizard/utils/typing_compat.py:199: error: Incompatible types in assignment (expression has type "Tuple[Any, Any]", variable has type "Tuple[Any, Type[_SpecialForm], Any, Any]")
dataclass_wizard/utils/typing_compat.py:200: error: Module has no attribute "_FinalTypingBase"
dataclass_wizard/utils/typing_compat.py:201: error: Module has no attribute "GenericMeta"; maybe "Generic"?
dataclass_wizard/utils/typing_compat.py:205: error: Module "typing_extensions" has no attribute "AnnotatedMeta"; maybe "Annotated"?
dataclass_wizard/utils/typing_compat.py:376: error: Module has no attribute "_eval_type"
dataclass_wizard/utils/typing_compat.py:386: error: "FREF" is a type variable and only valid in type context
dataclass_wizard/utils/type_conv.py:62: error: Argument 1 to "round" has incompatible type "Union[str, int, float, bool, None]"; expected "SupportsRound[Any]"
dataclass_wizard/utils/type_conv.py:65: error: Unsupported right operand type for in ("Union[str, int, float, bool, None]")
dataclass_wizard/utils/type_conv.py:66: error: Argument 1 to "float" has incompatible type "Union[str, int, float, bool, None]"; expected "Union[SupportsFloat, SupportsIndex, str, bytes]"
dataclass_wizard/utils/type_conv.py:208: error: Item "Number" of "Union[str, Number, datetime]" has no attribute "replace"
dataclass_wizard/utils/type_conv.py:208: error: Argument 1 to "replace" of "datetime" has incompatible type "str"; expected "int"
dataclass_wizard/utils/type_conv.py:208: error: Argument 2 to "replace" of "datetime" has incompatible type "str"; expected "int"
dataclass_wizard/utils/type_conv.py:299: error: Argument 1 to "replace" of "time" has incompatible type "str"; expected "int"
dataclass_wizard/utils/type_conv.py:299: error: Argument 2 to "replace" of "time" has incompatible type "str"; expected "int"
dataclass_wizard/utils/type_conv.py:345: error: Item "int" of "Union[str, int, timedelta]" has no attribute "replace"
dataclass_wizard/utils/type_conv.py:345: error: Item "timedelta" of "Union[str, int, timedelta]" has no attribute "replace"
dataclass_wizard/utils/type_conv.py:345: error: Item "float" of "Union[str, float, timedelta]" has no attribute "replace"
dataclass_wizard/utils/type_conv.py:345: error: Item "timedelta" of "Union[str, float, timedelta]" has no attribute "replace"
dataclass_wizard/utils/type_conv.py:345: error: Item "complex" of "Union[str, complex, timedelta]" has no attribute "replace"
dataclass_wizard/utils/type_conv.py:345: error: Item "timedelta" of "Union[str, complex, timedelta]" has no attribute "replace"
dataclass_wizard/utils/type_conv.py:346: error: Argument 1 to "float" has incompatible type "Union[str, int, timedelta]"; expected "Union[SupportsFloat, SupportsIndex, str, bytes]"
dataclass_wizard/utils/type_conv.py:346: error: Argument 1 to "float" has incompatible type "Union[str, float, timedelta]"; expected "Union[SupportsFloat, SupportsIndex, str, bytes]"
dataclass_wizard/utils/type_conv.py:346: error: Argument 1 to "float" has incompatible type "Union[str, complex, timedelta]"; expected "Union[SupportsFloat, SupportsIndex, str, bytes]"
dataclass_wizard/utils/type_conv.py:353: error: Incompatible types in assignment (expression has type "Union[str, int, timedelta]", variable has type "float")
dataclass_wizard/utils/type_conv.py:353: error: Incompatible types in assignment (expression has type "Union[str, float, timedelta]", variable has type "float")
dataclass_wizard/utils/type_conv.py:353: error: Incompatible types in assignment (expression has type "Union[str, complex, timedelta]", variable has type "float")
dataclass_wizard/property_wizard.py:193: error: Attribute function "default_factory" with type "Callable[[], _T]" does not accept self argument
dataclass_wizard/property_wizard.py:211: error: Value of type variable "FREF" of "eval_forward_ref_if_needed" cannot be "Optional[Type[Any]]"
dataclass_wizard/property_wizard.py:221: error: Argument 2 to "_default_from_generic_type" has incompatible type "Optional[Type[Any]]"; expected "Type[Any]"
dataclass_wizard/property_wizard.py:223: error: Argument 1 to "_default_from_type" has incompatible type "Optional[Type[Any]]"; expected "Type[Any]"
dataclass_wizard/property_wizard.py:241: error: Argument "default_factory" to "field" has incompatible type "Type[T]"; expected "Callable[[], Field[Any]]"
dataclass_wizard/property_wizard.py:243: error: Argument "default" to "field" has incompatible type "T"; expected "Field[Any]"
dataclass_wizard/property_wizard.py:265: error: Dict entry 0 has incompatible type "Optional[str]": "Type[T]"; expected "str": "Type[Any]"
dataclass_wizard/property_wizard.py:265: error: Argument 3 to "_process_field" has incompatible type "Optional[str]"; expected "str"
dataclass_wizard/property_wizard.py:268: error: Dict entry 0 has incompatible type "Optional[str]": "Type[T]"; expected "str": "Type[Any]"
dataclass_wizard/property_wizard.py:268: error: Argument 3 to "_default_from_annotation" has incompatible type "Optional[str]"; expected "str"
dataclass_wizard/property_wizard.py:311: error: Attribute function "default_factory" with type "Callable[[], _T]" does not accept self argument
dataclass_wizard/property_wizard.py:314: error: Attribute function "default_factory" with type "Callable[[], _T]" does not accept self argument
dataclass_wizard/models.py:3: error: Module "dataclasses" has no attribute "_create_fn"
dataclass_wizard/models.py:26: error: Variable "dataclass_wizard.bases.META" is not valid as a type
dataclass_wizard/models.py:26: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
dataclass_wizard/models.py:173: error: Type variable "dataclass_wizard.type_def.DT" is unbound
dataclass_wizard/models.py:173: note: (Hint: Use "Generic[DT]" or "Protocol[DT]" base class to bind "DT" inside a class)
dataclass_wizard/models.py:173: note: (Hint: Use "DT" in function signature to bind "DT" inside a function)
dataclass_wizard/abstractions.py:26: error: Name "META" is not defined
dataclass_wizard/abstractions.py:122: error: Type variable "dataclass_wizard.type_def.T" is unbound
dataclass_wizard/abstractions.py:122: note: (Hint: Use "Generic[T]" or "Protocol[T]" base class to bind "T" inside a class)
dataclass_wizard/abstractions.py:122: note: (Hint: Use "T" in function signature to bind "T" inside a function)
dataclass_wizard/class_helper.py:26: error: Name "META" is not defined
dataclass_wizard/class_helper.py:35: error: "DictWithLowerStore" expects no type arguments, but 2 given
dataclass_wizard/class_helper.py:59: error: Type variable "dataclass_wizard.bases.M" is unbound
dataclass_wizard/class_helper.py:59: note: (Hint: Use "Generic[M]" or "Protocol[M]" base class to bind "M" inside a class)
dataclass_wizard/class_helper.py:59: note: (Hint: Use "M" in function signature to bind "M" inside a function)
dataclass_wizard/class_helper.py:115: error: "DictWithLowerStore" expects no type arguments, but 2 given
dataclass_wizard/class_helper.py:129: error: "DictWithLowerStore" expects no type arguments, but 2 given
dataclass_wizard/parsers.py:26: error: Name "META" is not defined
dataclass_wizard/parsers.py:26: note: (Skipping most remaining errors due to unresolved imports or missing stubs; fix these first)
dataclass_wizard/parsers.py:79: error: Type variable "dataclass_wizard.type_def.M" is unbound
dataclass_wizard/parsers.py:79: note: (Hint: Use "Generic[M]" or "Protocol[M]" base class to bind "M" inside a class)
dataclass_wizard/parsers.py:79: note: (Hint: Use "M" in function signature to bind "M" inside a function)
dataclass_wizard/parsers.py:176: error: Type variable "dataclass_wizard.type_def.T" is unbound
dataclass_wizard/parsers.py:176: note: (Hint: Use "Generic[T]" or "Protocol[T]" base class to bind "T" inside a class)
dataclass_wizard/parsers.py:176: note: (Hint: Use "T" in function signature to bind "T" inside a function)
dataclass_wizard/parsers.py:266: error: Type variable "dataclass_wizard.type_def.LSQ" is unbound
dataclass_wizard/parsers.py:266: note: (Hint: Use "Generic[LSQ]" or "Protocol[LSQ]" base class to bind "LSQ" inside a class)
dataclass_wizard/parsers.py:266: note: (Hint: Use "LSQ" in function signature to bind "LSQ" inside a function)
dataclass_wizard/parsers.py:321: error: Type variable "dataclass_wizard.type_def.S" is unbound
dataclass_wizard/parsers.py:321: note: (Hint: Use "Generic[S]" or "Protocol[S]" base class to bind "S" inside a class)
dataclass_wizard/parsers.py:321: note: (Hint: Use "S" in function signature to bind "S" inside a function)
dataclass_wizard/parsers.py:424: error: Type variable "dataclass_wizard.type_def.S" is unbound
dataclass_wizard/parsers.py:424: note: (Hint: Use "Generic[S]" or "Protocol[S]" base class to bind "S" inside a class)
dataclass_wizard/parsers.py:424: note: (Hint: Use "S" in function signature to bind "S" inside a function)
dataclass_wizard/parsers.py:462: error: Type variable "dataclass_wizard.type_def.S" is unbound
dataclass_wizard/parsers.py:462: note: (Hint: Use "Generic[S]" or "Protocol[S]" base class to bind "S" inside a class)
dataclass_wizard/parsers.py:462: note: (Hint: Use "S" in function signature to bind "S" inside a function)
dataclass_wizard/parsers.py:488: error: Type variable "dataclass_wizard.type_def.M" is unbound
dataclass_wizard/parsers.py:488: note: (Hint: Use "Generic[M]" or "Protocol[M]" base class to bind "M" inside a class)
dataclass_wizard/parsers.py:488: note: (Hint: Use "M" in function signature to bind "M" inside a function)
dataclass_wizard/parsers.py:516: error: Type variable "dataclass_wizard.type_def.DD" is unbound
dataclass_wizard/parsers.py:516: note: (Hint: Use "Generic[DD]" or "Protocol[DD]" base class to bind "DD" inside a class)
dataclass_wizard/parsers.py:516: note: (Hint: Use "DD" in function signature to bind "DD" inside a function)
dataclass_wizard/parsers.py:540: error: Type variable "dataclass_wizard.type_def.S" is unbound
dataclass_wizard/parsers.py:540: note: (Hint: Use "Generic[S]" or "Protocol[S]" base class to bind "S" inside a class)
dataclass_wizard/parsers.py:540: note: (Hint: Use "S" in function signature to bind "S" inside a function)
dataclass_wizard/dumpers.py:13: error: Module "dataclasses" has no attribute "_is_dataclass_instance"
dataclass_wizard/dumpers.py:17: error: Module "typing" has no attribute "NamedTupleMeta"; maybe "NamedTuple"?
dataclass_wizard/loaders.py:6: error: Module "typing" has no attribute "NamedTupleMeta"; maybe "NamedTuple"?
dataclass_wizard/__init__.py:130: error: Library stubs not installed for "backports.datetime_fromisoformat" (or incompatible with Python 3.8)
dataclass_wizard/__init__.py:130: note: Hint: "python3 -m pip install types-backports"
dataclass_wizard/__init__.py:130: note: (or run "mypy --install-types" to install all missing stub packages)
dataclass_wizard/__init__.py:130: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
```

**Comment by AdiNar** [comment]
I will give it a try, at least to fix some of the type errors.

**Comment by rnag** [comment]
Should be closed/addressed by #64.

@tahouse Please re-open if you notice any further `mypy` issues or errors. Thanks!
